### PR TITLE
Add display features to MediaQuery

### PR DESCRIPTION
**Pull request by andreidiaconu**
This PR adds `DisplayFeature` to `MediaQuery` in the form of two new properties: `displayFeatures` and `hinge`. This is split from a larger PR for foldable support in order to make it easier to review (https://github.com/flutter/flutter/pull/77156). 

Issues that will be fixed by this PR:
- https://github.com/flutter/flutter/issues/49411

Depends on:
- Engine support: https://github.com/flutter/engine/pull/29447
- Framework integration of the engine support: https://github.com/flutter/flutter/pull/89511

PRs that depend on this:
- https://github.com/flutter/flutter/pull/92907
- https://github.com/flutter/flutter/pull/92909
- https://github.com/flutter/flutter/pull/92910

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by andreidiaconu** [comment]
Tests will not pass until the engine support lands but since the code is ready to be reviewed, I am marking this as ready for review.

File: packages/flutter/lib/src/widgets/media_query.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/media_query.dart:361]: nit (here and elsewhere): avoid "you", see https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#use-the-passive-voice-recommend-do-not-require

Maybe just say ", which lists the different types of display features."

File: packages/flutter/lib/src/widgets/media_query.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/media_query.dart:368]: nit: keep these to about 80chars, https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#prefer-a-maximum-line-length-of-80-characters

File: packages/flutter/lib/src/widgets/media_query.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/media_query.dart:374]: Why does the hinge get special treatment over other display features? What if a future phone introduces three (or more hinges)?

Comment (andreidiaconu) [reply packages/flutter/lib/src/widgets/media_query.dart:376]: The `hinge` is the only one of the three that forces apps to adapt. The `fold` does not stop pixels from rendering and the `cutout` is easily negotiated using SafeArea.
The initial reasoning was that having a nullable `hinge` property would make it easier for developers, but I now see it wasn't used in the rest of the PRs. 

Since the reasoning for having it is not rock solid, would you prefer if I remove it?

Comment (goderbauer) [reply packages/flutter/lib/src/widgets/media_query.dart:376]: I would just remove it then.

**Review by goderbauer — APPROVED** [review]
LGTM once tests pass (w/ engine commit)

Per our policy this will also need a second LGTM to land. I'll find a reviewer.

**Review by justinmc — APPROVED** [review]
LGTM 👍 

Just a docs nit and a comment about confirming List vs. Set.

File: packages/flutter/lib/src/widgets/media_query.dart
Comment (andreidiaconu) [reply packages/flutter/lib/src/widgets/media_query.dart:366]: The order does not matter but the data does travel as arrays inside the cpp engine. Would converting that to a set for usage in MediaQuery be ok?

Comment (andreidiaconu) [reply packages/flutter/lib/src/widgets/media_query.dart:366]: Also, while converting it to Set, I ran into this compile error in one of the tests: `The type of an element in a constant set can't override the '==' operator, but the type 'DisplayFeature' does.`
Does this mean that making it a Set would enable certain limitations in the future, when devs maybe want to init this as `const` (Can't really imagine a scenario outside of tests right now).

Comment (andreidiaconu) [reply packages/flutter/lib/src/widgets/media_query.dart:366]: One other solution to have the best of both worlds could be removing the `==` operator from DisplayFeature, but those are on the engine PR, not here. Do you feel strongly about having this be a Set?

**Comment by Piinks** [comment]
(Triage) I think this is blocked on 
- https://github.com/flutter/flutter/pull/93240
- https://github.com/flutter/engine/pull/29585

Which are blocked on https://github.com/flutter/plugins/pull/4502

@andreidiaconu if that's not correct, can you update?

**Comment by andreidiaconu** [comment]
> (Triage) I think this is blocked on
> 
> * [Reland 3: Display Features #93240](https://github.com/flutter/flutter/pull/93240)
> * [Reland 3: Display Features support engine#29585](https://github.com/flutter/engine/pull/29585)
> 
> Which are blocked on [flutter/plugins#4502](https://github.com/flutter/plugins/pull/4502)
> 
> @andreidiaconu if that's not correct, can you update?

That is correct.

File: packages/flutter/lib/src/widgets/media_query.dart
Comment (justinmc) [reply packages/flutter/lib/src/widgets/media_query.dart:366]: Sorry I missed these comments before.  I think it's ok to keep this as a List.  I think it's probably more important to override `==` than to use a Set here.  I didn't realize a const Set item couldn't have `==` overridden but it makes sense.

**Comment by andreidiaconu** [comment]
@blasten @Piinks I think this can be merged now.

**Comment by Piinks** [comment]
Hey @andreidiaconu can you update this with the latest from master? The other changes have rolled in, so the Google testing shard should pass with an update.

**Comment by andreidiaconu** [comment]
@Piinks Yes & done 👍

**Comment by Piinks** [comment]
So, I have checked the customer test out locally and ran it with your changes here and they all passed. I think an update will finally allow this to land. :) Thanks for your patience.

**Comment by goderbauer** [comment]
I re-ran the failing test. If that doesn't make it pass, you'll probably have to rebase this to the latest master.

**Comment by Piinks** [comment]
It is still failing after re-run, can you rebase once more?

**Comment by andreidiaconu** [comment]
Rebased

**Comment by Piinks** [comment]
@Hixie can you take a look at this rainbowmonkey failure?
I've seen it pop up on other PRs, but it usually goes away after rebasing. It does not appear to be related to this change.

**Comment by nt4f04uNd** [comment]
I recall that I once had it and it was solved when used `git rebase` instead, you might try it

**Comment by Hixie** [comment]
That was an unrelated issue due to rolling Dart, you should be able to resolve it by updating your PR to top of tree (i.e. rebasing as @nt4f04uNd says). You just got unlucky about what commit you started from.

**Comment by andreidiaconu** [comment]
Rebased this time around.

**Comment by Piinks** [comment]
That did it! Thanks all! #teamwork 😉

**Pull request merged**
